File: src/voice/panel.js
const { ActionRowBuilder, ButtonBuilder, ButtonStyle, PermissionFlagsBits } = require('discord.js');
const { embed } = require('../utils/embed');

const button = (id, label, style = ButtonStyle.Secondary) => new ButtonBuilder().setCustomId(`vc:${id}`).setLabel(label).setStyle(style);

function components() {
  return [
    new ActionRowBuilder().addComponents(button('lock', 'Lock'), button('unlock', 'Unlock'), button('hide', 'Hide'), button('unhide', 'Unhide')),
    new ActionRowBuilder().addComponents(button('rename', 'Rename'), button('limit', 'Limit'), button('bitrate', 'Bitrate')),
    new ActionRowBuilder().addComponents(button('permit', 'Permit'), button('reject', 'Reject'), button('transfer', 'Transfer'), button('claim', 'Claim')),
    new ActionRowBuilder().addComponents(button('info', 'Info'), button('delete', 'Delete', ButtonStyle.Danger)),
  ];
}

/** Read the channel's current state straight from its permission overwrites, so the panel is never out of sync. */
function state(channel) {
  const everyone = channel.permissionOverwrites.cache.get(channel.guild.id);
  return {
    locked: everyone?.deny.has(PermissionFlagsBits.Connect) ?? false,
    hidden: everyone?.deny.has(PermissionFlagsBits.ViewChannel) ?? false,
  };
}

function panelEmbed(channel, ownerId) {
  const { locked, hidden } = state(channel);
  return embed(
    [
      `Welcome to your channel, <@${ownerId}>. Use the buttons below to manage it.`,
      '',
      '**Lock / Unlock** — control who can join',
      '**Hide / Unhide** — control who can see it',
      '**Rename · Limit · Bitrate** — change channel settings',
      '**Permit / Reject** — allow or block specific users',
      '**Transfer** — hand ownership to someone else',
      '**Claim** — take over if the owner has left',
    ].join('\n'),
    {
      title: 'Voice controls',
      fields: [
        { name: 'Owner', value: `<@${ownerId}>`, inline: true },
        { name: 'Status', value: `${locked ? 'Locked' : 'Open'} · ${hidden ? 'Hidden' : 'Visible'}`, inline: true },
        { name: 'Limit', value: channel.userLimit ? `${channel.userLimit}` : 'None', inline: true },
      ],
    },
  );
}

module.exports = { components, panelEmbed, state };

File: bunny/voice/panel.py
"""The dashboard embed shown in each temporary voice channel's chat."""

from __future__ import annotations

import discord

from ..embeds import embed


def state(channel: discord.VoiceChannel) -> tuple[bool, bool]:
    """(locked, hidden), read straight from the permission overwrites so the panel is never out of sync."""
    overwrite = channel.overwrites_for(channel.guild.default_role)
    return overwrite.connect is False, overwrite.view_channel is False


def status_text(channel: discord.VoiceChannel) -> str:
    locked, hidden = state(channel)
    return f"{'Locked' if locked else 'Open'} · {'Hidden' if hidden else 'Visible'}"


def panel_embed(channel: discord.VoiceChannel, owner_id: int) -> discord.Embed:
    return embed(
        "\n".join(
            [
                f"Welcome to your channel, <@{owner_id}>. Use the buttons below to manage it.",
                "",
                "**Lock / Unlock**: control who can join",
                "**Hide / Unhide**: control who can see it",
                "**Rename · Limit · Bitrate**: change channel settings",
                "**Permit / Reject**: allow or block specific users",
                "**Transfer**: hand ownership to someone else",
                "**Claim**: take over if the owner has left",
            ]
        ),
        title="Voice controls",
        fields=[
            ("Owner", f"<@{owner_id}>", True),
            ("Status", status_text(channel), True),
            ("Limit", str(channel.user_limit) if channel.user_limit else "None", True),
        ],
    )
